engine/engine.go

```diff
@@ -0,0 +1,352 @@
+package engine
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/go-hclog"
+
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/gruntwork-io/go-commons/errors"
+	"github.com/gruntwork-io/terragrunt-engine-go/engine"
+	"github.com/gruntwork-io/terragrunt-engine-go/proto"
+	"github.com/gruntwork-io/terragrunt/options"
+	"github.com/gruntwork-io/terragrunt/util"
+	"github.com/hashicorp/go-plugin"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/anypb"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+const (
+	engineVersion     = 1
+	engineCookieKey   = "engine"
+	engineCookieValue = "terragrunt"
+)
+
+var (
+	engineClients = sync.Map{}
+)
+
+type ExecutionOptions struct {
+	TerragruntOptions *options.TerragruntOptions
+	CmdStdout         io.Writer
+	CmdStderr         io.Writer
+	WorkingDir        string
+	SuppressStdout    bool
+	AllocatePseudoTty bool
+	Command           string
+	Args              []string
+}
+
+type engineInstance struct {
+	terragruntEngine *proto.EngineClient
+	client           *plugin.Client
+	executionOptions *ExecutionOptions
+}
+
+// IsEngineEnabled returns true if the experimental engine is enabled.
+func IsEngineEnabled() bool {
+	switch strings.ToLower(os.Getenv("TG_EXPERIMENTAL_ENGINE")) {
```

Comment:
I would use a constant for "TG_EXPERIMENTAL_ENGINE":

``` go
const EnableExperimentalEngineEnvName = "TG_EXPERIMENTAL_ENGINE"
```
We will be able to reference this name from other places, which will reduce the likelihood that in the future it will be changed in one place and the old name will remain in others.